### PR TITLE
feat: add animated transition to frame on double-click.

### PR DESCRIPTION
**Pull request by jfree**
Adds an animated transition between frames upon double-click, using the Radiance animation library (https://github.com/kirill-grouchnikov/radiance).  To make this work, a small refactoring has been done to remove state from the `FrameGraphPainter` (specifically `flameGraphWidth` and `visibleWidth`) that can more easily be passed to the paint methods when called.

To test this PR, run the main app, load a JFR file and double-click on some frames.

### CODE REVIEW

**Comment by bric3** [comment]
Nice this works well. I wonder though if the external library could be _plugged in_ as most of its code is the `ZoomTarget` consumer.

I haven't looked at all the code yet. But I like the ZoomTarget notion and that the width is now specific to the canvas.

File: fireplace-swing/src/main/java/io/github/bric3/fireplace/flamegraph/FlameGraph.java [hunk 1]
```diff
@@ -325,14 +364,15 @@ public void mouseMoved(MouseEvent e) {
...
-                  ));
+                    .ifPresent(fgp -> fgp.hoverFrameAt(
+                            (Graphics2D) view.getGraphics(),
+                            canvas.getBounds(),
+                            point,
+                            frame -> {
+                                canvas.setToolTipText(frame);
+                                scrollPane.repaint();
+                            }
+                    ));
```
Comment (bric3) [inline fireplace-swing/src/main/java/io/github/bric3/fireplace/flamegraph/FlameGraph.java:375]: For reference might conflict with #34 https://github.com/bric3/fireplace/pull/34/files#diff-e367861376ff2c84840ac2b45bbf62c25e02fc13bbe3f5a44f6d5372bed5b3dbR329-R335

Let's merge #34 after this one as it's much simpler.

**Comment by bric3** [comment]
Enabling the debug mode `-Dfireplace.swing.debug=true` I noticed this triggers a lot of violations. This may cause some issues if this components gets embedded in an that checks these (like IJ).

```
EDT violation detected on component javax.swing.JScrollPane[,0,0,1000x475,invalid,layout=javax.swing.ScrollPaneLayout$UIResource,alignmentX=0.0,alignmentY=0.0,border=com.formdev.flatlaf.ui.FlatBorder@22284f87,flags=320,maximumSize=,minimumSize=,preferredSize=,columnHeader=,horizontalScrollBar=javax.swing.JScrollPane$ScrollBar[,3,462,984x10,layout=com.formdev.flatlaf.ui.FlatScrollBarUI,alignmentX=0.0,alignmentY=0.0,border=,flags=4194632,maximumSize=,minimumSize=,preferredSize=,blockIncrement=10,orientation=HORIZONTAL,unitIncrement=16],horizontalScrollBarPolicy=HORIZONTAL_SCROLLBAR_AS_NEEDED,lowerLeft=,lowerRight=,rowHeader=,upperLeft=,upperRight=,verticalScrollBar=javax.swing.JScrollPane$ScrollBar[,987,3,10x459,layout=com.formdev.flatlaf.ui.FlatScrollBarUI,alignmentX=0.0,alignmentY=0.0,border=,flags=4194632,maximumSize=,minimumSize=,preferredSize=,blockIncrement=10,orientation=VERTICAL,unitIncrement=16],verticalScrollBarPolicy=VERTICAL_SCROLLBAR_AS_NEEDED,viewport=javax.swing.JViewport[,3,3,984x459,invalid,layout=javax.swing.ViewportLayout,alignmentX=0.0,alignmentY=0.0,border=,flags=25165832,maximumSize=,minimumSize=,preferredSize=,isViewSizeSet=true,lastPaintPosition=java.awt.Point[x=-7705,y=-172],scrollUnderway=true],viewportBorder=]
    at java.desktop/javax.swing.JComponent.repaint(JComponent.java:4836)
    at java.desktop/javax.swing.JViewport.repaint(JViewport.java:1509)
    at java.desktop/java.awt.Component.repaint(Component.java:3436)
    at java.desktop/java.awt.Component.repaintParentIfNeeded(Component.java:2402)
    at java.desktop/java.awt.Component.reshape(Component.java:2390)
    at java.desktop/javax.swing.JComponent.reshape(JComponent.java:4241)
    at java.desktop/java.awt.Component.setBounds(Component.java:2326)
    at java.desktop/java.awt.Component.move(Component.java:2158)
    at java.desktop/java.awt.Component.setLocation(Component.java:2140)
    at java.desktop/javax.swing.JViewport.setViewPosition(JViewport.java:1240)
    at io.github.bric3.fireplace.flamegraph.FlameGraph$FlameGraphMouseInputListener$1.onTimelinePulse(FlameGraph.java:316)
    at org.pushingpixels.radiance.animation.api.Timeline$Chain.handlePulse(Timeline.java:298)
    at org.pushingpixels.radiance.animation.api.Timeline$Chain.onTimelinePulse(Timeline.java:315)
    at org.pushingpixels.radiance.animation.api.TimelineEngine.lambda$callbackCallTimelinePulse$3(TimelineEngine.java:505)
    at org.pushingpixels.radiance.animation.api.TimelineEngine$TimelineCallbackThread.run(TimelineEngine.java:182)

EDT violation detected on component io.github.bric3.fireplace.flamegraph.FlameGraph$FlameGraphCanvas[,-9825,-219,11108x108535,invalid,layout=java.awt.FlowLayout,alignmentX=0.0,alignmentY=0.0,border=,flags=9,maximumSize=,minimumSize=,preferredSize=]
    at java.desktop/javax.swing.JComponent.repaint(JComponent.java:4836)
    at java.desktop/java.awt.Component.repaint(Component.java:3393)
    at java.desktop/java.awt.Component.repaintParentIfNeeded(Component.java:2404)
    at java.desktop/java.awt.Component.reshape(Component.java:2390)
    at java.desktop/javax.swing.JComponent.reshape(JComponent.java:4241)
    at java.desktop/java.awt.Component.setBounds(Component.java:2326)
    at java.desktop/java.awt.Component.move(Component.java:2158)
    at java.desktop/java.awt.Component.setLocation(Component.java:2140)
    at java.desktop/javax.swing.JViewport.setViewPosition(JViewport.java:1240)
    at io.github.bric3.fireplace.flamegraph.FlameGraph$FlameGraphMouseInputListener$1.onTimelinePulse(FlameGraph.java:316)
    at org.pushingpixels.radiance.animation.api.Timeline$Chain.handlePulse(Timeline.java:298)
    at org.pushingpixels.radiance.animation.api.Timeline$Chain.onTimelinePulse(Timeline.java:315)
    at org.pushingpixels.radiance.animation.api.TimelineEngine.lambda$callbackCallTimelinePulse$3(TimelineEngine.java:505)
    at org.pushingpixels.radiance.animation.api.TimelineEngine$TimelineCallbackThread.run(TimelineEngine.java:182)
```

**Comment by jfree** [comment]
> I didn't see a toggle to enable or disable animations. (This might go along with plugability)

Now there is an attribute `animateZoomTransitions` on the `FlameGraphCanvas` which defaults to `true`.  This can be disabled programmatically, or by setting the System property `fireplace.zoom.animation.disabled` at program startup.

**Comment by bric3** [comment]
> I guess the animation library would cause trouble if you want to get this into JMC.  What kind of pluggability did you have in mind? 

Since all the animation code is contained as a consumer [here](https://github.com/bric3/fireplace/pull/37/files#diff-e367861376ff2c84840ac2b45bbf62c25e02fc13bbe3f5a44f6d5372bed5b3dbR287-R326), we can create a separate gradle module that extends the core flameGraph by configuring the consumer.

> Another alternative is I just roll-my-own timeline pulse generator with java.util.concurrent - I don’t think it would be that hard.

Possibly but I still think this could be made as a separate module.

Bu we can move extract this in another PR, to move quickly.

**Comment by kirill-grouchnikov** [comment]
Just happened to see this PR. Haven't looked closely at everything, but one thing jumped out - EDT violations. Radiance animation has a special annotation to mark callbacks to be dispatched on the EDT - `org.pushingpixels.radiance.animation.api.swing.RunOnEventDispatchThread`. There's also the base `EventDispatchThreadTimelineCallbackAdapter` that can be used without the explicit need to wrap stuff with `SwingUtilities.invokeLater`.

**Comment by jfree** [comment]
Thanks Kirill, I'm going to take a look at this.

**Pull request merged**